supernode/daemon/mgr/ha/etcd_tool.go

```diff
@@ -0,0 +1,146 @@
+	},error
+}
+
+func (etcd *etcdMgr)WathActiveChange(messageChannel chan string){
+	var watchStartRevision  int64
+	watcher := clientv3.NewWatcher(etcd.client)
+	// watch
+	watchChan := watcher.Watch(context.TODO(),"/lock/active",clientv3.WithRev(watchStartRevision))
+	for watchResp := range watchChan{
+		for _,event := range watchResp.Events{
+			switch event.Type {
+			case mvccpb.PUT:
+				//put means someone becomes a new active supernode
+				fmt.Println("change to:",string(event.Kv.Value),"Revision:",event.Kv.CreateRevision,event.Kv.ModRevision)
+				messageChannel<-"string(event.Kv.Value)"
+			case mvccpb.DELETE:
+				//deleted means the active supernode is off
+				fmt.Println("deleted, ","Revision:",event.Kv.ModRevision)
+				messageChannel<-"0.0.0.0"
+			}
+		}
+	}
+}
+
+func (etcd *etcdMgr)ObtainActiveInfo(key  string)(string,error){
+	kv := clientv3.NewKV(etcd.client)
+	var (
+		getRes *clientv3.GetResponse
+		err error
+		)
+	if getRes,err = kv.Get(context.TODO(),key);
+	 err !=nil{
+	 	log.Fatal(err)
+	 }
+	 var value  string
+	 for _,v := range getRes.Kvs {
+	 	value=string(v.Value)
+	 }
+	 return  value,err
+}
+func (etcd *etcdMgr)ActiveResureItsStatus(finished chan bool)bool{
+	for {
+		select {
+		case keepResp :=<-etcd.leaseKeepAliveRsp:
+			if keepResp == nil {
+				fmt.Println("lease renew fail ")
+				finished<-true
+				return false
+			}else{
+				//fmt.Println("lease renew success")
+			}
+		}
+	}
+}
+
+func (etcd *etcdMgr)TryBeActive(finished chan bool)(bool,string,error){
+	kv := clientv3.NewKV(etcd.client)
+	//make a lease to obtain a lock
+	lease:= clientv3.NewLease(etcd.client)
+	leaseResp,err := lease.Grant(context.TODO(),etcd.leaseTtl)
+	keepRespChan, kaerr := lease.KeepAlive(context.TODO(), leaseResp.ID)
+	etcd.leaseKeepAliveRsp=keepRespChan
+	if kaerr != nil {
+		log.Fatal(kaerr)
+	}
+	//go etcd.WatchLeaseRenew(finished)
+	etcd.leaseResp=leaseResp
+	//if the lock is available,get the lock
+	//else read the lock
+	txn:= kv.Txn(context.TODO())
+	txn.If(clientv3.Compare(clientv3.CreateRevision("/lock/active"),"=",0)).
+		Then(clientv3.OpPut("/lock/active",etcd.host_ip,clientv3.WithLease(leaseResp.ID))).
+		Else(clientv3.OpGet("/lock/active"))   //i will deal with it
+	txnResp,err := txn.Commit()
+	if err != nil{
+		log.Fatal(err)
+	}
+	if !txnResp.Succeeded{
+		fmt.Println("can not get lock")
+		_, err = lease.Revoke(context.TODO(), leaseResp.ID)
+		//active_supernode_ip<-string(txnResp.Responses[0].GetResponseRange().Kvs[0].Value)
+		return  false,string(txnResp.Responses[0].GetResponseRange().Kvs[0].Value),err
+	}else {
+		fmt.Println("get lock")
+		//active_supernode_ip<-etcd.ip
+		return true,etcd.host_ip,err
+	}
+}
+
+//Cancellation of lease
+func (etcd *etcdMgr)ActiveKillItself()bool{
+	_, err :=etcd.client.Revoke(context.TODO(),etcd.leaseResp.ID)
+	if err != nil {
+		fmt.Printf("Cancellation of lease fail\n")
+		return false
+	}
+	fmt.Printf("Cancellation of lease succsess\n")
+	return  true
+}
+
```

Comment:
We do not need to the following white spaces.

Reply:
done